File: android/app/src/main/assets/www/js/pdf-materiais-viatura.js
function tituloSecaoMateriais(doc, y, texto) {
  doc.setFont("helvetica", "bold");
  doc.setFontSize(11);
  doc.setTextColor(45, 106, 79);
  doc.text(String(texto), PDF_MARGEM, y);
  doc.setTextColor(0, 0, 0);
  return y + 7;
}

function desenharCabecalhoMateriais(doc, item) {
  const pageW = doc.internal.pageSize.getWidth();
  const center = pageW / 2;
  let y = PDF_MARGEM;
  const logo = typeof LOGO_BBA_DATA_URL !== "undefined" ? LOGO_BBA_DATA_URL : null;

  if (logo) {
    try {
      doc.addImage(logo, "PNG", center - 14, y, 28, 28);
      y += 32;
    } catch (e) {
      console.warn("Logo nao incluido no PDF:", e);
    }
  }

  doc.setFont("helvetica", "bold");
  doc.setFontSize(10);
  doc.setTextColor(27, 67, 50);
  [
    "SECRETARIA DE SEGURANÇA PÚBLICA",
    "CORPO DE BOMBEIROS MILITAR DO MARANHÃO",
    "BATALHÃO DE BOMBEIROS AMBIENTAL",
  ].forEach((linha) => {
    doc.text(linha, center, y, { align: "center" });
    y += 5;
  });

  doc.setFontSize(11);
  doc.text("CONFERÊNCIA DE MATERIAIS DA VIATURA", center, y + 2, { align: "center" });
  y += 10;

  doc.setDrawColor(45, 106, 79);
  doc.setLineWidth(0.5);
  doc.line(PDF_MARGEM, y, pageW - PDF_MARGEM, y);
  doc.setTextColor(0, 0, 0);
  return y + 8;
}

function textoViaturaPdf(v) {
  const ar = String(v.ar || "").trim();
  const marca = String(v.marca || "").trim();
  const modelo = String(v.modelo || "").trim();
  const placa = String(v.placa || "").trim();
  const nome = `${marca} ${modelo}`.trim();
  if (ar && nome) return placa ? `AR ${ar} — ${nome} · ${placa}` : `AR ${ar} — ${nome}`;
  return ar || placa || "—";
}

function nomeArquivoMateriais(item) {
  const data = String(item.data_servico || "conferencia").replace(/-/g, "");
  return `conferencia_materiais_${data}_${item.id || "0"}.pdf`;
}

function gerarPdfMateriais(item, opts) {
  const { jsPDF } = window.jspdf;
  const doc = new jsPDF({ unit: "mm", format: "a4" });
  let y = desenharCabecalhoMateriais(doc, item);

  const servico = [
    ["Data do serviço", dataExtensoMaterial(item.data_servico)],
    ["Dia da semana", item.dia_semana || "—"],
    ["Chefe de socorro", item.chefe_socorro || "—"],
  ];
  if (item.contato) servico.push(["Contato funcional", item.contato]);
  if (item.condutor) servico.push(["Motorista", item.condutor]);
  if (item.comandante) servico.push(["Comandante de guarnição", item.comandante]);
  if (item.oficial_dia) servico.push(["Oficial de dia", item.oficial_dia]);
  if (item.tipo_checklist) {
    servico.push([
      "Tipo de checklist",
      item.tipo_checklist === "foto" ? "Checklist com Foto" : "Checklist Manual",
    ]);
  }

  doc.autoTable({
    ...estiloTabelaPadrao(),
    head: [["Campo", "Valor"]],
    body: servico,
    startY: y,
    headStyles: { fillColor: [45, 106, 79] },
    columnStyles: { 0: { cellWidth: 52, fontStyle: "bold" }, 1: { cellWidth: 118 } },
  });
  y = doc.lastAutoTable.finalY + 6;

  const viaturas = item.viaturas || [];
  if (viaturas.length) {
    y = tituloSecaoMateriais(doc, y, "Viaturas");
    doc.autoTable({
      ...estiloTabelaPadrao(9),
      head: [["Viatura", "KM inicial", "Abastecimento"]],
      body: viaturas.map((v) => [
        textoViaturaPdf(v),
        v.km_inicial || v.km || "—",
        v.abastecimento || v.k7 || "—",
      ]),
      startY: y,
      headStyles: { fillColor: [45, 106, 79] },
      columnStyles: { 0: { cellWidth: 88 }, 1: { cellWidth: 38 }, 2: { cellWidth: 44 } },
    });
    y = doc.lastAutoTable.finalY + 6;
  }

  (item.secoes || []).forEach((secao) => {
    y = tituloSecaoMateriais(doc, y, secao.titulo || "Materiais");
    const body = (secao.itens || []).map((mat) => {
      const qtd = mat.status === "alterado" && mat.qtd_nova ? mat.qtd_nova : mat.qtd;
      const obs = [mat.obs, mat.status === "alterado" ? mat.obs_nova : ""].filter(Boolean).join(" · ");
      return [qtdLabelMaterial(qtd), mat.nome || "—", (mat.status || "pronto"), obs || "—"];
    });
    doc.autoTable({
      ...estiloTabelaPadrao(8),
      head: [["Qtd", "Material", "Status", "Observação"]],
      body,
      startY: y,
      headStyles: { fillColor: [45, 106, 79] },
      columnStyles: {
        0: { cellWidth: 14, halign: "center" },
        1: { cellWidth: 72 },
        2: { cellWidth: 24 },
        3: { cellWidth: 60 },
      },
    });
    y = doc.lastAutoTable.finalY + 6;
  });

  const assinaturas = item.assinaturas || {};
  if (assinaturas.saindo || assinaturas.entrando) {
    y = tituloSecaoMateriais(doc, y, "Passagem de serviço");
    if (assinaturas.saindo) {
      doc.setFont("helvetica", "bold");
      doc.setFontSize(10);
      doc.text(`Saindo: ${assinaturas.saindo.nome || "—"}`, PDF_MARGEM, y);
      y += 5;
      if (assinaturas.saindo.imagem) {
        try {
          const fmt = assinaturas.saindo.imagem.includes("image/png") ? "PNG" : "JPEG";
          doc.addImage(assinaturas.saindo.imagem, fmt, PDF_MARGEM, y, 70, 24);
          y += 28;
        } catch (e) {
          console.warn("Assinatura saindo nao incluida:", e);
        }
      }
    }
    if (assinaturas.entrando) {
      doc.setFont("helvetica", "bold");
      doc.text(`Entrando: ${assinaturas.entrando.nome || "—"}`, PDF_MARGEM, y);
      y += 5;
      if (assinaturas.entrando.imagem) {
        try {
          const fmt = assinaturas.entrando.imagem.includes("image/png") ? "PNG" : "JPEG";
          doc.addImage(assinaturas.entrando.imagem, fmt, PDF_MARGEM, y, 70, 24);
          y += 28;
        } catch (e) {
          console.warn("Assinatura entrando nao incluida:", e);
        }
      }
    }
  }

  const pageH = doc.internal.pageSize.getHeight();
  if (y + 10 > pageH - PDF_MARGEM) {
    doc.addPage();
    y = PDF_MARGEM;
  }
  doc.setFont("helvetica", "italic");
  doc.setFontSize(8);
  doc.setTextColor(120, 120, 120);
  doc.text(
    "Documento gerado automaticamente — Sistema BBA/CBMMA.",
    PDF_MARGEM,
    y + 4
  );

  const nomeArquivo = nomeArquivoMateriais(item);
  if (opts && opts.returnDoc) return { doc, nomeArquivo };
  salvarPdf(doc, nomeArquivo);
}

async function abrirPdfMateriais(id) {
  try {
    const item = obterConferenciaMateriais(id);
    if (!item) {
      alert("Conferência não encontrada.");
      return;
    }
    if (!window.jspdf) {
      alert("Biblioteca PDF não carregada.");
      return;
    }
    if (typeof garantirAutoTable === "function" && !garantirAutoTable()) {
      alert("Plugin de tabelas PDF não carregado.");
      return;
    }
    gerarPdfMateriais(item);
  } catch (e) {
    console.error("Erro ao gerar PDF:", e);
    alert("Não foi possível gerar o PDF.");
  }
}

async function compartilharPdfMateriais(id) {
  try {
    const item = obterConferenciaMateriais(id);
    if (!item) {
      alert("Conferência não encontrada.");
      return;
    }
    if (!window.jspdf) {
      alert("Biblioteca PDF não carregada.");
      return;
    }
    if (typeof garantirAutoTable === "function" && !garantirAutoTable()) {
      alert("Plugin de tabelas PDF não carregado.");
      return;
    }
    const { doc, nomeArquivo } = gerarPdfMateriais(item, { returnDoc: true });
    const blob = doc.output("blob");
    const titulo = `Conferência ${dataExtensoMaterial(item.data_servico)}`;
    const file = new File([blob], nomeArquivo, { type: "application/pdf" });
    if (navigator.share && navigator.canShare && navigator.canShare({ files: [file] })) {
      await navigator.share({ files: [file], title: titulo });
      return;
    }
    salvarPdf(doc, nomeArquivo);
  } catch (e) {
    if (e && e.name === "AbortError") return;
    console.error("Erro ao compartilhar PDF:", e);
    alert("Não foi possível compartilhar o PDF.");
  }
}

function bindAcoesMateriais(root) {
  const scope = root || document;
  scope.querySelectorAll(".btn-acao-pdf-materiais").forEach((btn) => {
    btn.addEventListener("click", async (e) => {
      e.preventDefault();
      e.stopPropagation();
      if (btn.disabled) return;
      const label = btn.textContent;
      btn.disabled = true;
      btn.textContent = "Gerando…";
      try {
        await abrirPdfMateriais(btn.dataset.id);
      } finally {
        btn.disabled = false;
        btn.textContent = label;
      }
    });
  });
  scope.querySelectorAll(".btn-acao-share-materiais").forEach((btn) => {
    btn.addEventListener("click", async (e) => {
      e.preventDefault();
      e.stopPropagation();
      if (btn.disabled) return;
      const label = btn.textContent;
      btn.disabled = true;
      btn.textContent = "Preparando…";
      try {
        await compartilharPdfMateriais(btn.dataset.id);
      } finally {
        btn.disabled = false;
        btn.textContent = label;
      }
    });
  });
}

File: app/pdf_materiais_viatura.py
"""Geração de PDF — conferência de materiais da viatura."""

import base64
import os
import re
from io import BytesIO

from reportlab.lib import colors
from reportlab.lib.enums import TA_CENTER
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import cm
from reportlab.platypus import Image, Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle

from app.materiais_viatura import (
    abastecimento_viatura_conferencia,
    data_extenso,
    km_viatura_conferencia,
    qtd_label,
    texto_viatura_cadastro,
)
from app.pdf_report import LOGO_BBA_PATH


def _table_style_materiais(font_size=9):
    return TableStyle(
        [
            ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#2d6a4f")),
            ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
            ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
            ("GRID", (0, 0), (-1, -1), 0.5, colors.lightgrey),
            ("TOPPADDING", (0, 0), (-1, -1), 5),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 5),
            ("FONTSIZE", (0, 0), (-1, -1), font_size),
            ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, colors.HexColor("#f6f9f7")]),
            ("VALIGN", (0, 0), (-1, -1), "TOP"),
        ]
    )


def _imagem_data_url(data_url: str, max_w=5 * cm, max_h=2.2 * cm):
    if not data_url or not str(data_url).startswith("data:image"):
        return None
    match = re.match(r"data:image/(png|jpeg|jpg);base64,(.+)", data_url, re.I)
    if not match:
        return None
    try:
        raw = base64.b64decode(match.group(2))
        return Image(BytesIO(raw), width=max_w, height=max_h, kind="proportional")
    except Exception:
        return None


def gerar_pdf_conferencia_materiais(conferencia: dict) -> bytes:
    buffer = BytesIO()
    doc = SimpleDocTemplate(
        buffer,
        pagesize=A4,
        rightMargin=2 * cm,
        leftMargin=2 * cm,
        topMargin=1.8 * cm,
        bottomMargin=1.8 * cm,
    )

    styles = getSampleStyleSheet()
    header_line = ParagraphStyle(
        "CabLine",
        parent=styles["Normal"],
        fontSize=10,
        alignment=TA_CENTER,
        fontName="Helvetica-Bold",
        textColor=colors.HexColor("#1b4332"),
        spaceAfter=3,
        leading=12,
    )
    title_line = ParagraphStyle(
        "CabTitle",
        parent=header_line,
        fontSize=11,
        spaceBefore=6,
        spaceAfter=10,
    )
    heading_style = ParagraphStyle(
        "Section",
        parent=styles["Heading2"],
        fontSize=11,
        textColor=colors.HexColor("#2d6a4f"),
        spaceBefore=10,
        spaceAfter=6,
    )
    body_style = styles["Normal"]
    subtitle_style = ParagraphStyle(
        "Subtitle",
        parent=styles["Normal"],
        fontSize=9,
        textColor=colors.grey,
        spaceAfter=8,
    )

    story = []
    data_label = data_extenso(conferencia.get("data_servico") or "")

    if os.path.exists(LOGO_BBA_PATH):
        logo = Image(LOGO_BBA_PATH, width=2.8 * cm, height=2.8 * cm, kind="proportional")
        logo_table = Table([[logo]], colWidths=[16 * cm])
        logo_table.setStyle(
            TableStyle(
                [
                    ("ALIGN", (0, 0), (-1, -1), "CENTER"),
                    ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
                    ("BOTTOMPADDING", (0, 0), (-1, -1), 4),
                ]
            )
        )
        story.append(logo_table)

    for linha in (
        "SECRETARIA DE SEGURANÇA PÚBLICA",
        "CORPO DE BOMBEIROS MILITAR DO MARANHÃO",
        "BATALHÃO DE BOMBEIROS AMBIENTAL",
    ):
        story.append(Paragraph(linha, header_line))

    story.append(Paragraph("CONFERÊNCIA DE MATERIAIS DA VIATURA", title_line))

    resumo = [
        [Paragraph("<b>Data do serviço</b>", body_style), data_label],
        [Paragraph("<b>Dia da semana</b>", body_style), conferencia.get("dia_semana") or "—"],
        [Paragraph("<b>Chefe de socorro</b>", body_style), conferencia.get("chefe_socorro") or "—"],
    ]
    for rotulo, chave in (
        ("Contato funcional", "contato"),
        ("Motorista", "condutor"),
        ("Comandante de guarnição", "comandante"),
        ("Oficial de dia", "oficial_dia"),
    ):
        valor = (conferencia.get(chave) or "").strip()
        if valor:
            resumo.append([Paragraph(f"<b>{rotulo}</b>", body_style), valor])

    tipo = (conferencia.get("tipo_checklist") or "").strip()
    if tipo:
        label_tipo = "Checklist com Foto" if tipo == "foto" else "Checklist Manual"
        resumo.append([Paragraph("<b>Tipo de checklist</b>", body_style), label_tipo])

    t_resumo = Table(resumo, colWidths=[5.5 * cm, 10.5 * cm])
    t_resumo.setStyle(
        TableStyle(
            [
                ("BACKGROUND", (0, 0), (-1, -1), colors.HexColor("#f6f9f7")),
                ("BOX", (0, 0), (-1, -1), 1, colors.HexColor("#2d6a4f")),
                ("GRID", (0, 0), (-1, -1), 0.25, colors.HexColor("#c8e6c9")),
                ("TOPPADDING", (0, 0), (-1, -1), 6),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 6),
                ("LEFTPADDING", (0, 0), (-1, -1), 8),
                ("RIGHTPADDING", (0, 0), (-1, -1), 8),
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
            ]
        )
    )
    story.append(t_resumo)
    story.append(Spacer(1, 0.35 * cm))

    viaturas = conferencia.get("viaturas") or []
    if viaturas:
        story.append(Paragraph("Viaturas", heading_style))
        linhas = [["Viatura", "KM inicial", "Abastecimento"]]
        for v in viaturas:
            linhas.append(
                [
                    texto_viatura_cadastro(v),
                    km_viatura_conferencia(v),
                    abastecimento_viatura_conferencia(v),
                ]
            )
        t_vtr = Table(linhas, colWidths=[8 * cm, 4 * cm, 4 * cm])
        t_vtr.setStyle(_table_style_materiais())
        story.append(t_vtr)
        story.append(Spacer(1, 0.25 * cm))

    for secao in conferencia.get("secoes") or []:
        story.append(Paragraph(secao.get("titulo") or "Materiais", heading_style))
        linhas = [["Qtd", "Material", "Status", "Observação"]]
        for item in secao.get("itens") or []:
            qtd = item.get("qtd_nova") if item.get("status") == "alterado" and item.get("qtd_nova") else item.get("qtd")
            obs_parts = []
            if item.get("obs"):
                obs_parts.append(item["obs"])
            if item.get("status") == "alterado" and item.get("obs_nova"):
                obs_parts.append(item["obs_nova"])
            linhas.append(
                [
                    qtd_label(qtd),
                    item.get("nome") or "—",
                    (item.get("status") or "pronto").capitalize(),
                    " · ".join(obs_parts) or "—",
                ]
            )
        t = Table(linhas, colWidths=[1.4 * cm, 7.2 * cm, 2.4 * cm, 5 * cm])
        t.setStyle(_table_style_materiais(8))
        story.append(t)
        story.append(Spacer(1, 0.15 * cm))

    assinaturas = conferencia.get("assinaturas") or {}
    if assinaturas.get("saindo") or assinaturas.get("entrando"):
        story.append(Paragraph("Passagem de serviço", heading_style))
        blocos = []
        for rotulo, chave in (("Militar saindo", "saindo"), ("Militar entrando", "entrando")):
            dados = assinaturas.get(chave)
            if not dados:
                continue
            img = _imagem_data_url(dados.get("imagem") or "")
            celulas = [[Paragraph(f"<b>{rotulo}</b><br/>{dados.get('nome') or '—'}", body_style)]]
            if img:
                celulas.append([img])
            blocos.append(celulas)

        if blocos:
            cols = len(blocos)
            row_nome = [bloco[0][0] for bloco in blocos]
            rows = [row_nome]
            if any(len(b) > 1 for b in blocos):
                rows.append([bloco[1][0] if len(bloco) > 1 else "" for bloco in blocos])
            col_w = 16 * cm / cols
            t_assin = Table(rows, colWidths=[col_w] * cols)
            t_assin.setStyle(
                TableStyle(
                    [
                        ("BOX", (0, 0), (-1, -1), 1, colors.HexColor("#2d6a4f")),
                        ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#f6f9f7")),
                        ("ALIGN", (0, 0), (-1, -1), "CENTER"),
                        ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
                        ("TOPPADDING", (0, 0), (-1, -1), 8),
                        ("BOTTOMPADDING", (0, 0), (-1, -1), 8),
                    ]
                )
            )
            story.append(t_assin)

    fotos = conferencia.get("fotos") or []
    if fotos:
        story.append(Spacer(1, 0.3 * cm))
        story.append(Paragraph("Fotos da conferência", heading_style))
        for idx, src in enumerate(fotos[:3], start=1):
            img = _imagem_data_url(src, max_w=12 * cm, max_h=8 * cm)
            if img:
                story.append(Paragraph(f"Foto {idx}", subtitle_style))
                story.append(img)
                story.append(Spacer(1, 0.2 * cm))

    story.append(Spacer(1, 0.5 * cm))
    story.append(
        Paragraph(
            "<i>Documento gerado automaticamente — Sistema BBA/CBMMA.</i>",
            subtitle_style,
        )
    )

    doc.build(story)
    buffer.seek(0)
    return buffer.read()
